myst_nb/cache.py

```diff
@@ -0,0 +1,192 @@
+"""
+Implements integration of jupyter-cache
+"""
+import os
+import nbformat as nbf
+from nbclient import execute
+from pathlib import Path
+
+from sphinx.util import logging
+from sphinx.util.osutil import ensuredir
+
+from jupyter_cache.cache.db import NbCacheRecord
+from jupyter_cache import get_cache
+from jupyter_cache.executors import load_executor
+
+logger = logging.getLogger(__name__)
+
+
+def execution_cache(app, env, added, changed, removed, path_cache=None):
+    """
+    If cacheing is required, stages and executes the added or modified notebooks,
+    and caches then for further use.
+    """
+    jupyter_cache = False
+    nb_list = added.union(
+        changed
+    )  # all the added and changed notebooks should be operated on.
+    if env.config["jupyter_execute_notebooks"]:
+        jupyter_cache = env.config["jupyter_cache"]
+
+    if jupyter_cache is True:
+        path_cache = path_cache or Path(env.outdir).parent.joinpath(".jupyter_cache")
+        app.env.path_cache = (
+            path_cache  # TODO: is there a better way to make it accessible?
+        )
+
+        stage_and_execute(env, nb_list, path_cache)
+
+    return nb_list  # TODO: can also compare timestamps for inputs outputs
+
+
+def stage_and_execute(env, nb_list, path_cache):
+    pk_list = None
+    do_run = False
+
+    if path_cache:
+        try:
+            from jupyter_cache.cache.main import JupyterCacheBase  # noqa: F401
+        except ImportError:
+            logger.error(
+                (
+                    "Using caching functionality requires that "
+                    "jupyter_cache is uninstalled. Install it first."
+                )
+            )
+
+        cache_base = get_cache(path_cache)
+
+        if "jupyter_notebook_force_run" in env.config:
+            do_run = env.config["jupyter_notebook_force_run"]
+
+        for nb in nb_list:
+            # excludes the file with patterns given in execution_excludepatterns
+            # conf variable from executing, like index.rst
+            exclude_file = [x in nb for x in env.config["execution_excludepatterns"]]
+            if True in exclude_file:
+                continue
+
+            has_outputs = False
+
+            if "/" in nb:  # nb includes the path to notebook
```

Comment:
doesn't doc2path also operate on files with `/` in the name? (e.g. if the file is in a sub-directory of the docs?

Reply:
Yes, this was a stupid mistake. This is also the reason why execution in docs was failing with `file not found` which you mentioned in your `first thoughts from using it` comment. I guess `if "." in nb` should work, since `added` and `removed` file lists created by sphinx does not have extensions in the list.